### DISCUSSION TITLE
ESP8266/ESP32 Pin Interrupt False Positives

### DISCUSSION THREAD

**Discussion by PsuFan** [micropython/micropython#9223] [Hardware & Peripherals]
I was wondering if this had a simple explanation or solution because I run into this a lot lately. Recently I have a 15 ft CAT5 cable with:

1 pair for switched 3.3 relay / 1 pair ground for relay
1 pair gpio switch / 1 pair ground for switch

Set the switch to input, pull up, with an interrupt trigger FALLING. When I turn on the relay i get 10,000 interrupt triggers on the switch that isn't connected. I also have a switch on a 2 inch wire and that interrupt is fine, does not get triggered. I haven't checked the voltage on this specific setup but I've checked with a setup with a 50 ft switch, no relay with the same issue and the voltage never goes under 3v. Replaced out with MCP23017 gpio extender and it works fine. Also tried using an external 1K pull up which did nothing.

**Comment by jimmo** [comment]
Sorry @PsuFan I'm having a bit of a hard time understanding your description of the circuits. Can you draw a diagram maybe?

I think what you're saying is that for a given pair, both conductors are connected. i.e. you're using a CAT5 cable as a 4-conductor cable, where each pair is a single conductor?

i.e.

```
blue-  --\          /-- relay1+
blue+ ----~~~~~~~~~~--- relay1+
orange- --\         /-- relay1-
orange+ ---~~~~~~~~~--- relay1-
brown+ --\          /-- switch1+
brown- ---~~~~~~~~~~--- switch1+
green+ --\          /-- switch1-
green- ---~~~~~~~~~~--- switch1-
```

Or are you saying that you have two relays and two switches?

```
blue-  ---------------- relay1+
blue+ ----------------- relay1+
orange- --------------- relay1-
orange+ --------------- relay2-
brown+ ---------------- switch1+
brown- ---------------- switch2+
green+ ---------------- switch1-
green- ---------------- switch2-
```

Neither of these particularly make sense for using CAT5. The relay turning on is very likely to induce a current in the switch lines. You'd be much better off with current return on the same pair.

**Reply by robert-hh** [reply]
The relay will create a LOT OF NOISE when switching. So this is a not so well wiring choice. At least you should use the wires from one pair of signal and GND. Then the shielding is a little bit better. So use 1 pair for the relay, one pair for the switch and connect all other wires to GND. And use a pull-up resistor of 220Ohm for the switch, with a 100nF capacitor at the pull-up Vcc side to GND. 
Maybe that helps. Better use shielded cables with a shield at each pair and the shield connected to GND.

**Reply by PsuFan** [reply]
Hey jimmo, the first circuit, 1 relay / 1 switch.

@robert-hh jeez 220Ohm seems very strong. What does noise have to do with it? As long as the voltage is above ~1.25 or whatever why would it trigger the falling interrupt? I also have this issue with long wiring (~50ft) and no relay (sometimes run with RFID though). Capacitor before or after the resistor?

ESP Pin---220Ohm---100nF---GND

ESP Pin---100nF---220Ohm---GND

Didn't think about induced current on the switched lines... why doesn't it affect the short 2 inch switch?

**Reply by jimmo** [reply]
What you've essentially built is an accidental antenna that transmits from the relay line to the switch line. i.e. turning on the relay induces a current in the switch line.

The strength of the pull-up is what the induced current has to "fight against" to toggle the pin.

The capacitor acts as a low-pass filter - this is standard practice on any button/switch circuit to provide "de-bouncing". i.e. you're filtering out the rapid changes in voltage in this line (either caused by switch contacts bouncing or in this case by the induced field).

**Reply by jimmo** [reply]
Remember there's no such thing as a sharp edge in electrical signals. Your pin turning on the relay actually causes a (decaying) oscillation in current due to the inductance of the wire and the coil (this is known as ringing, and it's essentially a low pass filtered version of the step function). It's this oscillation that you're picking up in the switch line.

**Reply by jimmo** [reply]
I just remembered I have some photos that show a very exaggerated version of this from a demo I was doing for someone last year.

This is a picture of a square wave on channel 1 going via a coax cable, and the same signal going via two (couple of metres) regular wires (you can see the green and white wires) into channel 2.
https://photos.app.goo.gl/2QqzDnqBGFVXvLAMA
(If you look closely you can also see the phase delay from the difference in wire length between the two channels)

There's no sinusoidal output from the signal generator here, this is purely an artefact of the wire.

This case is different to yours because it's going into a very high impedance whereas yours is going into a relay coil, but you get the idea... It's easy to make an accidental oscillator, and an oscillating current in a wire that's the right length is a radio transmitter. (And a wire of the same length right next to it is a receive antenna).

**Comment by robert-hh** [comment]
The low impedance serves for two purposes:
1. Lower the level of cross-coupled noise from the relay. the capacity of the cat cable is about 50pF/m. So at 15 ft it's~ 250pf, at 50 ft it's ~800pf. The low resistor with the capacitor will then lowers the noise.
2. It increases the rising slope of the signal when your switch opens. The ESP8266 and ESP32 create many irq triggers on a slow slope. And noise on the wire makes it worse. Other chips have a Schmitt-Trigger characterisic on their input. Then, this slow slope noise does not happen. 
3. And a long wire will pick up noise from mains anyhow. A low impedance load will lower that noise.

In addition to that, there will be contact bouncing when the switch is closed, causing multiple triggers as well.

See below a Oscilliscope of a ESP32 and a triangle signal at the input (green) and the response created by a pin IRQ handler, triggering on both rising/falling edge.
![triangle_resp_1000](https://user-images.githubusercontent.com/12476868/188689959-d77afe05-74c1-4797-a1d3-86eded9de70c.jpg)

The  diagram below shows what is read back on an input with a slow slope (green) and the value seen (yellow). The upper part shows the full window, the lower a zoom into the transition area. Device MIMXRT1020.

![transient_jitter_read_DR](https://user-images.githubusercontent.com/12476868/188691823-3c4f27e2-20b4-445a-8f70-94200542320b.png)

**Reply by PsuFan** [reply]
Thanks for all the info and time guys! I understand the antenna reference somewhat. What I don't understand is why the antenna would be detecting FALLING under ~1.25v. I could understand if it raised the voltage and triggered a rising when you turn on the relay, but how is it sucking the voltage out. Also I don't understand why I can swap out the ESP gpio with a gpio extender (MCP23017) which I believe uses 100K pull-ups that don't have the false positives (on the stand alone 50ft line). I cant really use debouncing methods in some cases because I connect to RFID readers which send data on 2ms @ 5v and I dont use logic level converters lol. I have limited access to oscilloscopes but have not been able to see any of these voltage drops causing the interrupt. Will a 100nF cap slope cause issues for 2ms pulsing data? Its wiegand 0 & 1 data lines, always high falling representing a 0 or 1 depending on line.
